### ISSUE TITLE
Optimistic update

### ISSUE THREAD

**Issue by ptbrowne** [cozy/cozy-client#378]
When saving a document, it can be interesting for some usecase when updating a document to not wait the response from the server to update the store. In our case in Banks, we had a performance problem when toggling balance panels.

**Comment by cedricmessiant** [comment]
I already used Apollo's optimistic update, it was mostly to improve UX. It's a great feature and it would be nice to have it in `cozy-client`

For those who don't what this is: https://blog.apollographql.com/tutorial-graphql-mutations-optimistic-ui-and-store-updates-f7b6b66bf0e2

**Comment by drazik** [comment]
I agree that it's a nice feature to have in cozy-client. But it seems it can be very difficult to implement it nicely to handle every cases. In banks, we need the « level 0 »: we optimistically update the UI, but if the underlying mutation failed, we don't care so much. Because if the user collapses a panel and see it's expanded when he refreshes the page, it's not a big failure.

A next level would be to be able to update the UI in response to a failed mutation (show an error message; undo the optimistic update; whatever). I didn't think of other « levels », but it certainly exists other ones.

What I want to say is that maybe we can first implement the « level 0 » and use it as an opt-in, consciously knowing that it's not perfect, if the entire perfect feature is too big.

**Comment by ptbrowne** [comment]
> What I want to say is that maybe we can first implement the « level 0 » and use it as an opt-in, consciously knowing that it's not perfect, if the entire perfect feature is too big.

👍

**Comment by edas** [comment]
Je vois trois cas : 
* On met à jour les données locales, on lance la mise à jour distante et on oublie
* On met à jour les données locales, on lance la mise à jour distante, on veut être prévenu s'il y a eu un problème
* On met à jour les données locales, on lance la mise à jour distante, on veut mettre à jour de nouveau les données avec la réponse distante une fois qu'elle est là


à priori : 
* `save(document, { optimistic: true })` (premier cas)
* `save(document, { optimistic: function(response) { … } })` où le callback peut répondre UPDATE_LOCAL (met à jour les données locales avec la réponse obtenue, troisième cas) ou NOUPDATE (ne fait rien, je m'occupe de tout, second cas)
